alg/teca_temporal_reduction.cxx

```diff
@@ -0,0 +1,1083 @@
+//       }
+//    }
+}
+
+// --------------------------------------------------------------------------
+template <typename T>
+int minimum(int device_id, T *p_out_array, T *p_out_valid, const T *p_in_array,
+    const T *p_in_valid, unsigned long n_elem)
+{
+    // determine kernel launch parameters
+    int n_blocks = 0;
+    dim3 block_grid;
+    dim3 thread_grid;
+    if (teca_cuda_util::partition_thread_blocks(device_id,
+        n_elem, 8, block_grid, n_blocks, thread_grid))
+    {
+        TECA_ERROR("Failed to partition thread blocks")
+        return -1;
+    }
+
+    // launch the finalize kernel
+    cudaError_t ierr = cudaSuccess;
+    minimum<<<block_grid,thread_grid>>>(p_out_array, p_out_valid, p_in_array, p_in_valid, n_elem);
+    if ((ierr = cudaGetLastError()) != cudaSuccess)
+    {
+        TECA_ERROR("Failed to launch the finalize CUDA kernel"
+            << cudaGetErrorString(ierr))
+        return -1;
+    }
+    return 0;
+}
+
+// --------------------------------------------------------------------------
+template <typename T>
+__global__
+void finalize(T *p_out_array, const T *p_out_valid, double fill_value, unsigned long n_elem)
+{
+    unsigned long i = teca_cuda_util::thread_id_to_array_index();
+
+    if (i >= n_elem)
+        return;
+
+    if (!p_out_valid[i])
+        p_out_array[i] = fill_value;
+}
+
+// --------------------------------------------------------------------------
+template <typename T>
+int finalize(int device_id, T *p_out_array, const T *p_out_valid, double fill_value,
+    unsigned long n_elem)
+{
+    // determine kernel launch parameters
+    int n_blocks = 0;
+    dim3 block_grid;
+    dim3 thread_grid;
+    if (teca_cuda_util::partition_thread_blocks(device_id,
+        n_elem, 8, block_grid, n_blocks, thread_grid))
+    {
+        TECA_ERROR("Failed to partition thread blocks")
+        return -1;
+    }
+
+    // launch the finalize kernel
+    cudaError_t ierr = cudaSuccess;
+    finalize<<<block_grid,thread_grid>>>(p_out_array, p_out_valid, fill_value, n_elem);
+    if ((ierr = cudaGetLastError()) != cudaSuccess)
+    {
+        TECA_ERROR("Failed to launch the finalize CUDA kernel"
+            << cudaGetErrorString(ierr))
+        return -1;
+    }
+    return 0;
+}
+}
+#endif
+
+// --------------------------------------------------------------------------
+void minimum_operator::update(
+    int device_id,
+    p_teca_variant_array &out_array,
+    p_teca_variant_array &out_valid,
+    const_p_teca_variant_array &in_array,
+    const_p_teca_variant_array &in_valid)
+{
+#if defined(TECA_HAS_CUDA)
+    // set the CUDA device to run on
+    cudaError_t ierr = cudaSuccess;
+    if ((ierr = cudaSetDevice(device_id)) != cudaSuccess)
+    {
+        TECA_ERROR("Failed to set the CUDA device to " << device_id
+            << ". " << cudaGetErrorString(ierr))
+//        return -1;
+    }
+#endif
+
+    unsigned long n_elem = out_array->size();
+
+    TEMPLATE_DISPATCH_FP_SI(
```

Comment:
this version of the dispatch macro excludes unsigned integers, but do we need to exclude them here?